src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -301,7 +343,7 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
 
             var endpoint = new MatcherEndpoint(
                 next => invokerDelegate,
-                RoutePatternFactory.Parse(template, defaults, constraints: null),
+                RoutePatternFactory.Pattern(patternRawText, defaults, constraints: null, segments),
```

Comment:
If MvcEndpointInfo does this during its construction, we can avoid this overload here and can use the one which just takes in segments.